File: src/main/java/info/skyblond/archivedag/security/JwtTokenManager.java
package info.skyblond.archivedag.security;

import info.skyblond.archivedag.config.JwtProperties;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.security.Key;
import java.util.Base64;
import java.util.Date;
import java.util.Objects;

@Slf4j
@Component
public class JwtTokenManager {
    private final JwtProperties jwtProperties;
    private final Key signingKey;

    public JwtTokenManager(JwtProperties jwtProperties) {
        this.jwtProperties = jwtProperties;
        if (jwtProperties.getSecret() == null) {
            log.info("No JWT secret found, generating one...");
            this.signingKey = Keys.secretKeyFor(jwtProperties.getSignatureAlgorithm());
            log.info("Generated JWT secret key: {}", Base64.getEncoder().encodeToString(this.signingKey.getEncoded()));
        } else {
            this.signingKey = Keys.hmacShaKeyFor(Base64.getDecoder().decode(jwtProperties.getSecret()));
        }
    }

    public String generateToken(UserDetails userDetails) {
        long currentTime = System.currentTimeMillis();
        long duration = this.jwtProperties.getExpireInUnit().toMillis(this.jwtProperties.getExpireInDuration());

        return Jwts.builder()
                .setSubject(userDetails.getUsername())
                .setIssuedAt(new Date(currentTime))
                .setNotBefore(new Date(currentTime))
                .setExpiration(new Date(currentTime + duration))
                .signWith(this.signingKey, this.jwtProperties.getSignatureAlgorithm())
                .compact();
    }

    public String getUsernameFromToken(String token) {
        Claims claims = Jwts.parserBuilder()
                .setSigningKey(this.signingKey)
                .build()
                .parseClaimsJws(token)
                .getBody();
        log.info("Get claims: {}", claims);
        Date now = new Date();
        if (claims.getNotBefore() != null && now.before(claims.getNotBefore())) {
            // now ... ntb
            return null;
        }
        // Here we reject token with no expiration
        if (Objects.requireNonNull(claims.getExpiration()).before(now)) {
            // exp ... now
            return null;
        }
        return claims.getSubject();
    }

}

File: backend/src/main/kotlin/info/skyblond/archivedag/arudaz/security/JwtTokenManager.kt
package info.skyblond.archivedag.arudaz.security

import info.skyblond.archivedag.arudaz.config.JwtProperties
import io.jsonwebtoken.Claims
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.slf4j.LoggerFactory
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey

@Component
class JwtTokenManager(private val jwtProperties: JwtProperties) {
    private val logger = LoggerFactory.getLogger(JwtTokenManager::class.java)
    private var signingKey: SecretKey? = null

    init {
        if (jwtProperties.secret == null) {
            logger.warn("No JWT secret found, generating one...")
            signingKey = Keys.secretKeyFor(jwtProperties.signatureAlgorithm)
            logger.warn(
                "Generated JWT secret key: {}",
                Base64.getEncoder().encodeToString(signingKey!!.encoded)
            )
        } else {
            signingKey = Keys.hmacShaKeyFor(Base64.getDecoder().decode(jwtProperties.secret))
        }
    }

    fun generateToken(userDetails: UserDetails): String {
        val currentTime = System.currentTimeMillis()
        val duration = jwtProperties.expireInUnit.toMillis(jwtProperties.expireInDuration)
        return Jwts.builder()
            .setSubject(userDetails.username)
            .setIssuedAt(Date(currentTime))
            .setNotBefore(Date(currentTime))
            .setExpiration(Date(currentTime + duration))
            .signWith(signingKey, jwtProperties.signatureAlgorithm)
            .compact()
    }

    fun getUsernameFromToken(token: String?): String? {
        val claims: Claims = try {
            Jwts.parserBuilder()
                .setSigningKey(signingKey)
                .build()
                .parseClaimsJws(token)
                .body
        } catch (e: JwtException) {
            logger.debug("Error when decoding JWT token", e)
            return null
        }
        return claims.subject
    }
}
